pkg/nodeagent/controller/healthcheck/kubelet.go

```diff
@@ -0,0 +1,264 @@
+	// getAddresses is a func which returns a slice of ip addresses on this node
+	getAddresses          func() ([]net.Addr, error)
+	kubeletHealthEndpoint string
+}
+
+// NewKubeletHealthChecker create a instance of a kubelet health check.
+func NewKubeletHealthChecker(client client.Client, clock clock.Clock, dbus dbus.DBus, recorder record.EventRecorder, getAddresses func() ([]net.Addr, error)) *KubeletHealthChecker {
+	return &KubeletHealthChecker{
+		client:                  client,
+		dbus:                    dbus,
+		Clock:                   clock,
+		recorder:                recorder,
+		getAddresses:            getAddresses,
+		KubeletReadinessToggles: []time.Time{},
+		kubeletHealthEndpoint:   defaultKubeletHealthEndpoint,
+	}
+}
+
+// Name returns the name of this health check.
+func (*KubeletHealthChecker) Name() string {
+	return "kubelet"
+}
+
+// HasLastInternalIP returns true if the node.InternalIP was stored.
+// Exported for testing.
+func (k *KubeletHealthChecker) HasLastInternalIP() bool {
+	return k.lastInternalIP.IsValid()
+}
+
+// SetKubeletHealthEndpoint set the kubeletHealthEndpoint.
+// Exported for testing.
+func (k *KubeletHealthChecker) SetKubeletHealthEndpoint(kubeletHealthEndpoint string) {
+	k.kubeletHealthEndpoint = kubeletHealthEndpoint
+}
+
+// Check performs the actual health check for the kubelet.
+func (k *KubeletHealthChecker) Check(ctx context.Context, node *corev1.Node) error {
+	log := logf.FromContext(ctx).WithName("kubelet")
+
+	// This mimics the old behavior of the kubelet-health-monitor which checks for
+	// to many NotReady->Ready toggles in a certain time and reboots the node in such a case.
+	if isNodeReady(node) && !k.NodeReady {
+		needsReboot := k.ToggleKubeletState()
+		log.Info("Kubelet became Ready", "readinessChanges", len(k.KubeletReadinessToggles), "timespan", toggleTimeSpan)
+		if needsReboot {
+			log.Info("Kubelet toggled between NotReady and Ready too often. Rebooting the node now")
+			k.recorder.Eventf(node, corev1.EventTypeWarning, "kubelet", "Kubelet toggled between NotReady and Ready at least %d times in a %s time window. Rebooting the node now", maxToggles, toggleTimeSpan)
+			err := k.dbus.Reboot()
+			if err != nil {
+				k.RevertToggleKubeletState()
+				k.recorder.Event(node, corev1.EventTypeWarning, "kubelet", "Rebooting the node failed")
+				return fmt.Errorf("rebooting the node failed %w", err)
+			}
+		}
+	}
+	k.NodeReady = isNodeReady(node)
+
+	err := k.ensureNodeInternalIP(ctx, node)
+	if err != nil {
+		return err
+	}
+
+	httpClient := &http.Client{
+		Timeout: 10 * time.Second,
+	}
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, k.kubeletHealthEndpoint, nil)
+	if err != nil {
+		log.Error(err, "Creating request to kubelet health endpoint failed")
+		return err
+	}
+	response, err := httpClient.Do(request)
+	if err != nil {
+		log.Error(err, "HTTP request to kubelet health endpoint failed")
+	}
+	if err == nil && response.StatusCode == http.StatusOK {
```

Comment:
Do we want to distinguish between a connection timeout and other errors?

Reply:
I think that it is sufficient to have the error message in the event. That should help in failure analysis.